### PR TITLE
Fix handling of visibility modifiers

### PR DESCRIPTION
**Pull request by mvz**
Fixes both #793 and #792.

### CODE REVIEW

**Comment by troessner** [comment]
Mein lieber Scholli!

Looking forward to reviewing this one!

**Comment by mvz** [comment]
Ready for review!

**Comment by troessner** [comment]
>> Ready for review!

Already penciled this bad mother in for my train ride to work tomorrow morning ;)

**Comment by mvz** [comment]
> Already penciled this bad mother in for my train ride to work tomorrow morning ;)

Enjoy!

File: lib/reek/context_builder.rb [hunk 1]
```diff
@@ -133,20 +148,12 @@ def process_defs(exp)
...
-      method_name = exp.method_name
-      if exp.visibility_modifier?
-        element.track_visibility(method_name, exp.arg_names)
-      elsif exp.attribute_writer?
-        exp.args.each do |arg|
-          append_new_context(Context::AttributeContext, arg, exp)
-        end
-      else
-        append_new_context(Context::SendContext, exp, method_name)
+      if current_context.module_context?
```
Comment (mvz) [inline lib/reek/context_builder.rb]: This is a kind of type checking in disguise and I'd love to handle this differently. Ideas welcome.

File: lib/reek/context/attribute_context.rb [hunk 2]
```diff
@@ -15,6 +15,15 @@ def full_comment
         send_expression.full_comment || ''
       end
 
+      def instance_method?
+        true
+      end
+
+      # Was this method defined with an instance method-like syntax?
```
Comment (troessner) [inline lib/reek/context/attribute_context.rb]: This is overwritten depending on the context, right? Can we somehow document in one place what we mean with "instance-like syntax" and then drop this comment here?

File: lib/reek/context/code_context.rb [hunk 3]
```diff
@@ -16,13 +15,16 @@ module Context
...
     class CodeContext
       extend Forwardable
       delegate each_node: :exp
       delegate %i(name type) => :exp
-      delegate %i(visibility visibility= non_public_visibility?) => :visibility_tracker
 
-      attr_reader :children, :context, :exp, :statement_counter, :visibility_tracker
+      attr_accessor :visibility
+      attr_reader :children, :context, :exp
+      attr_reader :statement_counter
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Why the additional attr_reader? Why not put it above?

File: lib/reek/context/code_context.rb [hunk 4]
```diff
@@ -16,13 +15,16 @@ module Context
...
     # :reek:TooManyInstanceVariables: { max_instance_variables: 8 }
+    # :reek:Attribute
     class CodeContext
       extend Forwardable
       delegate each_node: :exp
       delegate %i(name type) => :exp
-      delegate %i(visibility visibility= non_public_visibility?) => :visibility_tracker
 
-      attr_reader :children, :context, :exp, :statement_counter, :visibility_tracker
+      attr_accessor :visibility
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: So we are back to ....just a value. Using an object and composition + delegation feels cleaner to me. But maybe we dont need that anymore! I'll just keep on reading ;)

File: lib/reek/context/code_context.rb [hunk 5]
```diff
@@ -60,12 +62,13 @@ class CodeContext
...
       def initialize(context, exp)
-        @context            = context
         @exp                = exp
         @children           = []
-        @visibility_tracker = VisibilityTracker.new
         @statement_counter  = StatementCounter.new
         @refs               = AST::ObjectRefs.new
+        @visibility         = :public
+
+        @context = context.append_child_context(self) if context
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Can we put a helpful comment in here when context can be nil? Those things always trip me up in our codebase. I then always pull out the debugger to have a lightbulb moment but I think we should rather document it ;)

File: lib/reek/context/ghost_context.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
+require_relative 'code_context'
+require_relative 'singleton_method_context'
+
+module Reek
+  module Context
+    # Semi-transparent context to represent a metaclass while building the
```
Comment (troessner) [inline lib/reek/context/ghost_context.rb]: Yes, yes, yes! Thats exactly what we need.

**Comment by troessner** [comment]
Ok, that took me longer than expected. My train ride wasnt enough, I'm still reviewing, I'll finish this the next days, promised.

File: lib/reek/context_builder.rb (hunk 1 above)
Comment (tuexss) [inline lib/reek/context_builder.rb]: I think this began when you created two different methods with name `handle_` and including the type in the method name. If you can find a better naming for what these two methods do, you could either find a single abstraction that fits for both contexts, or you can find two different methods that apply in both contexts, but maybe with no effects for some invocations.

File: lib/reek/context_builder.rb [hunk 7]
```diff
@@ -465,8 +473,26 @@ def inside_new_context(klass, exp)
...
-        element.append_child_context new_context
+      klass.new(current_context, *args)
+    end
+
+    def handle_send_for_modules(exp)
+      method_name = exp.method_name
+      arg_names = exp.arg_names
+      current_context.track_visibility(method_name, arg_names)
+      current_context.track_singleton_visibility(method_name, arg_names)
+      register_attributes(exp) if exp.attribute_writer?
```
Comment (tuexss) [inline lib/reek/context_builder.rb]: How about moving the check for `exp.attribute_writer?` into the method `register_attributes` ?

File: lib/reek/context_builder.rb (hunk 1 above)
Comment (mvz) [inline lib/reek/context_builder.rb]: Actually, it began as a case statement over the type of `current_context`. I now think I should have left it like that to make the type checking at least explicit for later improvement.

File: lib/reek/context/ghost_context.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    # context tree. This context will not be part of the resulting tree, but
+    # will track context and visibility separately while building is in
+    # progress.
+    class GhostContext
+      attr_reader :children
+      private_attr_reader :visibility_tracker
+
+      def initialize(parent, _exp)
+        @context = parent
```
Comment (troessner) [inline lib/reek/context/ghost_context.rb]: How about starting to correct one of the biggest naming errors in our code base and call this "@parent"? Or would you rather prefer to do this in a separate PR?

Comment (mvz) [inline lib/reek/context/ghost_context.rb]: Good idea. I can include this in this PR.

Comment (troessner) [inline lib/reek/context/ghost_context.rb]: Great, then lets rename that across /contexts!

File: lib/reek/ast/sexp_extensions/methods.rb [hunk 9]
```diff
@@ -36,28 +36,9 @@ def body_nodes(types, ignoring = [])
...
-        # class << self syntax.
-        #
-        # To check for this we check if the parent node is of type :sclass.
-        #
-        # @return [Boolean]
-        def singleton_method_via_class_self_notation?
-          return unless parent
-          parent.type == :sclass
-        end
-      end
```
Comment (troessner) [inline lib/reek/ast/sexp_extensions/methods.rb]: :+1 for removing this from our sexp extensions

File: lib/reek/ast/sexp_extensions/send.rb [hunk 10]
```diff
@@ -38,10 +37,6 @@ def object_creation_call?
           method_name == :new
         end
 
-        def visibility_modifier?
-          VISIBILITY_MODIFIERS.include?(method_name)
-        end
-
```
Comment (troessner) [inline lib/reek/ast/sexp_extensions/send.rb]: :+1: for this as well. 
The more we put into `contexts` the better.

File: lib/reek/context/code_context.rb [hunk 11]
```diff
@@ -144,16 +143,35 @@ def config_for(detector_class)
...
+      def singleton_method?
+        false
+      end
+
+      def instance_method?
+        false
+      end
+
+      def defined_as_instance_method?
+        false
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Shouldn't all of those methods not rather raise `NotImplementedError` since we're applying the template pattern here and this is the mother base?

File: lib/reek/context/attribute_context.rb [hunk 12]
```diff
@@ -15,6 +15,15 @@ def full_comment
         send_expression.full_comment || ''
       end
 
+      def instance_method?
+        true
+      end
```
Comment (troessner) [inline lib/reek/context/attribute_context.rb]: Detected MethodEnvy Smell! See what I did there? I just invented a new smell type, cause this is how we roll!
In other news, I have a problem with the gazillion of overwritten methods. I know that we're using the template pattern here but it feels wrong since we're talking about predicate methods here. And we **know** what they return.

Long story short: Feels like we're mixing up "code" with "data" here.
How about using a map we reference via class instance variable in each context like this:

```Ruby
class AttributeContext < ...
  @yolo = {  # Trying to come up with a good name here. "method_properties"?
    instance_method: true,
    # ....
  }
```

wdyt?

File: lib/reek/context/code_context.rb (hunk 4 above)
Comment (troessner) [inline lib/reek/context/code_context.rb]: Ok, from what I gather reading is that you are using a `visibility` attribute on `CodeContext` that then the `VisibilityTracker` in all inherited contexts uses, right?
Personally I find this confusing. Can't we just put all-the-things in `VisibilityTracker`?

File: lib/reek/ast/sexp_extensions/send.rb (hunk 10 above)
Comment (troessner) [inline lib/reek/ast/sexp_extensions/send.rb]: To be more precise with: 

>> The more we put into contexts the better.

I believe our SexpExtensions should only **know** what you can see with the bare eye when looking at the corresponding Sexp. 
Everything that is slightly more complex should be put into `contexts` I believe.

File: lib/reek/context/code_context.rb [hunk 13]
```diff
@@ -95,16 +98,12 @@ def each(&block)
 
       alias_method :parent, :context
 
-      # Register a child context. The child's parent context should be equal to
-      # the current context.
-      #
-      # This makes the current context responsible for setting the child's
-      # visibility.
+      # Register a child context. Used by a child context to register itself with its parent.
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: But now the comment and the method name have diverged. 
Comment says "used by a child context to register itself".
Method name says "append_child_context". Shouldn't it then be "register_self"? And why do we pass in the "child"?
This is super-confusing.

File: lib/reek/context/code_context.rb (hunk 11 above)
Comment (troessner) [inline lib/reek/context/code_context.rb]: And additionally I don't think we should do it like that but rather use a map, please see my comment [here](https://github.com/troessner/reek/pull/796/files#diff-e4ff3ae7e7f4c263b7ab67ca5bb23e1fR20).

File: lib/reek/context/code_context.rb [hunk 14]
```diff
@@ -144,16 +143,35 @@ def config_for(detector_class)
...
+      def instance_method?
+        false
+      end
+
+      def defined_as_instance_method?
+        false
+      end
+
+      def apply_current_visibility(current_visibility)
+        return unless defined_as_instance_method?
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: This line is in `CodeContext` but it seems to be specific to MethodContexts, isn't it?

File: lib/reek/context/ghost_context.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        real_parent = @context.append_child_context(child)
+        visibility_tracker.set_child_visibility(child)
+        real_parent
+      end
+
+      # Return the correct class for child method contexts (representing nodes
+      # of type `:def`). For GhostContext, this is the class that represents
+      # singleton methods.
+      def method_context_class
+        SingletonMethodContext
```
Comment (troessner) [inline lib/reek/context/ghost_context.rb]: I would put this in a map as well.

File: lib/reek/context/ghost_context.rb [hunk 16]
```diff
@@ -0,0 +1,62 @@
+require_relative 'code_context'
+require_relative 'singleton_method_context'
+
+module Reek
+  module Context
+    # Semi-transparent context to represent a metaclass while building the
+    # context tree. This context will not be part of the resulting tree, but
+    # will track context and visibility separately while building is in
+    # progress.
+    class GhostContext
```
Comment (troessner) [inline lib/reek/context/ghost_context.rb]: Even though it is not a real `Context`, shouldn't it still inherit from `CodeContext`? Right now this class is just dangling in a sea of contexts that all inherit from `CodeContext`....
Additionally, you require `CodeContext` in the first line of the file but do not use it anywhere, so right now I'm not sure what the intention is here :(

**Comment by troessner** [comment]
@mvz I reviewed approximately 68% of this PR. I'll stop now to give you some time to actually respond to all my gazillion comments ;)

File: lib/reek/context/code_context.rb (hunk 4 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: I like the separation of the property 'visibility', and the thing that tracks the visibility. In fact, only methods and attributes have a visibility, while only modules and classes need a tracker. Ideally, I would therefore push down both the visibility tracker and visibility to the contexts that actually need them. That should clear up the confusion as well.

File: lib/reek/context/code_context.rb (hunk 3 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: It seemed nicer because statement_counter is a tool while the other two are properties. I'm not super attached to this change. Ideally, I would not even expose statement_counter, but that is for another PR.

File: lib/reek/context/code_context.rb (hunk 5 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: Renaming context to parent will make this clearer I think.

File: lib/reek/context/code_context.rb (hunk 13 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: I'll update this to explain it better. In short, the child is supposed to call this method on its parent.

File: lib/reek/context/code_context.rb (hunk 11 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: Some are a substitute for type checking and should just go. The others, I'm not sure yet but should probably only be defined where relevant.

File: lib/reek/context/code_context.rb (hunk 14 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: MethodContexts and AttributContexts, yes. I'll see if I can push this down or into a module.

File: lib/reek/context/ghost_context.rb (hunk 16 above)
Comment (mvz) [inline lib/reek/context/ghost_context.rb]: I thought it needed CodeContext, but then it turned out most of the methods that get used need a different implementation. I'll revisit this decision.

File: lib/reek/context_builder.rb (hunk 7 above)
Comment (mvz) [inline lib/reek/context_builder.rb]: That makes sense.

File: lib/reek/context/attribute_context.rb (hunk 12 above)
Comment (mvz) [inline lib/reek/context/attribute_context.rb]: I kind of see your point but I'm not sure how we would expose these values outside of the class, which is where they are used mostly.

**Comment by mvz** [comment]
@troessner awesome, thanks. I'll revisit the code sometime this week I think.

File: lib/reek/context/code_context.rb (hunk 5 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: I have extracted the creation of parent-child relations out of the initializer again.

**Comment by mvz** [comment]
@troessner I've fixed and/or commented on most of your points. Care to continue the review?

**Comment by troessner** [comment]
>> @troessner I've fixed and/or commented on most of your points. Care to continue the review?

I will! Just gimme a day or two, I'm firing from all cylinders right now!

File: lib/reek/context/attribute_context.rb (hunk 12 above)
Comment (troessner) [inline lib/reek/context/attribute_context.rb]: How about:

```Ruby
class AttributeContext
  @method_properties = {
    instance_method: true,
    singleton_method: false
  }
  @method_properties.keys.each do |predicate|
    define_singleton_method "#{predicate}?" do
      @method_properties.fetch predicate
    end
  end
end
```

?

That is super-hard to read, completely overengineered but looks like we know what the hell we are doing here!

Ok, on second thought let's go with your solution. #no-yolo

File: lib/reek/context/attribute_context.rb [hunk 17]
```diff
@@ -15,6 +19,14 @@ def full_comment
...
+
+      def apply_current_visibility(current_visibility)
```
Comment (troessner) [inline lib/reek/context/attribute_context.rb]: Since this method now basically boiled down to a setter, wouldn't it make moar sense to just use "attr_writer"?
Because right now the name is so pretentious, that it implies that there is a lot more happening than just a setter.

File: lib/reek/context/code_context.rb [hunk 18]
```diff
@@ -20,14 +19,14 @@ class CodeContext
...
 
-      attr_reader :children, :context, :exp, :statement_counter, :visibility_tracker
+      attr_reader :children, :parent, :exp, :statement_counter
+
       private_attr_reader :refs
 
       # Initializes a new CodeContext.
       #
-      # @param context [CodeContext, nil] The parent context
+      # @param parent [CodeContext, nil] The parent context
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Party time! There is it! The baddest, most confusing attribute name in `Reek`s history, awesome that we change this now.
Second baddest name is "SourceCode#description" btw.

File: lib/reek/context/code_context.rb [hunk 19]
```diff
@@ -136,24 +132,30 @@ def matches?(candidates)
       end
 
       def full_name
-        exp.full_name(context ? context.full_name : '')
+        exp.full_name(parent ? parent.full_name : '')
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Yep, way clearer already. Before that i would **always** think "what the hell is context here again?!", scroll up, down, left and right and the be like "ah, yeah, its just the friggin parent" :(

File: lib/reek/context/code_context.rb [hunk 20]
```diff
@@ -136,24 +132,30 @@ def matches?(candidates)
...
+      def singleton_method?
+        false
+      end
+
+      def instance_method?
+        false
+      end
+
+      def apply_current_visibility(_current_visibility)
+        # Nothing to do by default
```
Comment (troessner) [inline lib/reek/context/code_context.rb]: Ok, so we can't just use a standard setter like I suggested before. But can we still just name this method "visibility="?

File: lib/reek/context/ghost_context.rb (hunk 15 above)
Comment (troessner) [inline lib/reek/context/ghost_context.rb]: Ok, we can scratch that comment, see above.

File: lib/reek/context/method_context.rb [hunk 21]
```diff
@@ -35,7 +42,24 @@ def default_assignments
...
+
+      def apply_current_visibility(current_visibility)
+        self.visibility = current_visibility
+      end
+
+      def module_function?
+        visibility == :module_function
+      end
+
+      # @return [Boolean] If the visibility is public or not.
```
Comment (troessner) [inline lib/reek/context/method_context.rb]: As much as I love yard doc, I think we can skip comments like this. I mean, the method name literally says this already and it uses the predicate notation as well...

File: lib/reek/context_builder.rb [hunk 22]
```diff
@@ -16,16 +18,16 @@ module Reek
   # counting. Ideally `ContextBuilder` would only build up the context tree and leave the
   # statement and reference counting to the contexts.
   #
-  # :reek:TooManyMethods: { max_methods: 27 }
+  # :reek:TooManyMethods: { max_methods: 30 }
   # :reek:UnusedPrivateMethod: { exclude: [ !ruby/regexp /process_/ ] }
   class ContextBuilder
     attr_reader :context_tree
-    private_attr_accessor :element
+    private_attr_accessor :current_context
```
Comment (troessner) [inline lib/reek/context_builder.rb]: Oh, wait, I forgot "element" for my hitlist of bad naming in `Reek` :laughing: 
Love it that we fix this now as well.

**Comment by troessner** [comment]
Well, let me say that again, great job!
I did review it again and left a couple of minor comments, nothing big.
From my POV we can get this ready to merge - let's address my last comments and then squash this one!

File: lib/reek/context/code_context.rb (hunk 20 above)
Comment (mvz) [inline lib/reek/context/code_context.rb]: I tried this just now and it doesn't work, since the logic for setting the visibility is different depending on whether the current visibility is used or the visibility is set explicitly using e.g. `private` with arguments.

Comment (troessner) [inline lib/reek/context/code_context.rb]: Yes, ( I realized that below as well, see my other comment) but we can at least name it like that, can't we?
The pompous "apply_current_visibility" always makes me think "wow, there must be quite some hard sh*t going on" only then to realize that it is .... just a conditional setter.

Comment (mvz) [inline lib/reek/context/code_context.rb]: > we can at least name it like that, can't we?

I'm afraid we can't, because there's other logic that calls `.visibility=` directly that needs it to be a regular setter.

Comment (troessner) [inline lib/reek/context/code_context.rb]: Ok, lets just leave it like it is ;)

**Comment by troessner** [comment]
Merging this one, highly awesome pull request that addresses a lot issues!
